pkg/controller/certificaterequest/cloudflare.go

```diff
@@ -49,59 +49,84 @@ type CloudflareResponse struct {
 	CD        bool                 `json:"CD"`
 	Questions []CloudflareQuestion `json:"Question"`
 	Answers   []CloudflareAnswer   `json:"Answer"`
+	Authority []CloudflareAnswer   `json:"Authority"`
 }
 
-// VerifyDnsResourceRecordUpdate is used to export the verifyDnsResourceRecordUpdate
-// result back to the caller.
+// VerifyDnsResourceRecordUpdate verifies the presence of a TXT record with Cloudflare DNS.
 func VerifyDnsResourceRecordUpdate(reqLogger logr.Logger, fqdn string, txtValue string) bool {
-	var attempt int
-	var success bool
-
-	attempt = 1
-	// After checking attempt count is < then maxAttemptsForDnsPropogationCheck,
-	// verifyDnsResourceRecordUpdate will validate the challange with the presence
-	// of the DNS record with cloudflare. If validated, returns true.
-	for attempt < maxAttemptsForDnsPropogationCheck {
-		reqLogger.Info(fmt.Sprintf("attempt %v to verify resource record %v has been updated with value %v", attempt, fqdn, txtValue))
-		success = verifyDnsResourceRecordUpdate(reqLogger, fqdn, txtValue)
-		if success {
-			return success
+	var response *CloudflareResponse
+	var attempt int = 1
+
+	for attempt < maxAttemptsForDnsPropagationCheck {
+		var err error
+
+		// Sleep before querying Cloudflare DNS.  If the previous attempt returned
+		// a negative cache result, honor its TTL (within reason).  Otherwise wait
+		// for a predetermined duration.
+		sleepDuration := waitTimePeriodDnsPropagationCheck
+		if response != nil &&
+			dnsRCode(response.Status) == dnsRCodeNameError &&
+			len(response.Authority) > 0 {
+			ttl := response.Authority[0].TTL
+			reqLogger.Info(fmt.Sprintf("got a negative cache response with a TTL of %v seconds", ttl))
+			// maxNegativeCacheTTL determines what is "reasonable".
+			// If the SOA TTL exceeds this, give up immediately.
+			if sleepDuration > maxNegativeCacheTTL {
+				reqLogger.Info("negative cache TTL is too long; giving up")
+				return false
+			}
+			// If the TTL is shorter than the default wait time, disregard.
+			if ttl > sleepDuration {
+				sleepDuration = ttl
```

Comment:
Should we maybe add 5 extra seconds just to make sure thing things have expired?

Reply:
Good idea  :+1: